src/algebra/group/unique_prods.lean

```diff
@@ -0,0 +1,213 @@
+of `B`.  We call the formalization this property `unique_prods`.  Since the condition requires no
+property of the group operation, we define it for a Type simply satisfying `has_mul`.  We also
+introduce the analogous "additive" companion, `unique_sums` and link the two so that `to_additive`
+converts `unique_prods` into `unique_sums`.
+
+Here you can see several examples of Types that have `unique_sums/prods`
+(`apply_instance` uses `covariants.to_unique_prods` and `covariants.to_unique_sums`).
+```lean
+import data.real.basic
+
+example : unique_sums ℕ   := by apply_instance
+example : unique_sums ℕ+  := by apply_instance
+example : unique_sums ℤ   := by apply_instance
+example : unique_sums ℚ   := by apply_instance
+example : unique_sums ℝ   := by apply_instance
+example : unique_prods ℕ+ := by apply_instance
+```
+
+A Type that does not have `unique_prods`.
+```lean
+example : ¬ unique_prods ℕ :=
+begin
+  rintros ⟨h⟩,
+  refine not_not.mpr (h (finset.singleton_nonempty 0) (finset.insert_nonempty 0 {1})) _,
+  suffices : (∃ (x : ℕ), (x = 0 ∨ x = 1) ∧ ¬x = 0) ∧ ∃ (x : ℕ), (x = 0 ∨ x = 1) ∧ ¬x = 1,
+  { simpa [unique_mul] },
+  exact ⟨⟨1, by simp⟩, ⟨0, by simp⟩⟩,
+end
+```
+
+A Type that does not have `unique_sums`.
+```lean
+import data.zmod.basic
+
+example (n : ℕ) (n2 : 2 ≤ n): ¬ unique_sums (zmod n) :=
+begin
+  haveI : fintype (zmod n) := @zmod.fintype n ⟨(zero_lt_two.trans_le n2).ne'⟩,
+  haveI : nontrivial (zmod n) := char_p.nontrivial_of_char_ne_one (one_lt_two.trans_le n2).ne',
+  rintros ⟨h⟩,
+  refine not_not.mpr (h finset.univ_nonempty finset.univ_nonempty) _,
+  suffices : ∀ (x y : zmod n), ∃ (x' y' : zmod n), x' + y' = x + y ∧ (x' = x → ¬y' = y),
+  { simpa [unique_add] },
+  exact λ x y, ⟨x - 1, y + 1, sub_add_add_cancel _ _ _, by simp⟩,
+end
+```
+-/
+
+/--  Let `G` be a Type with multiplication, let `A B : finset G` be finite subsets and
+let `a0 b0 : G` be two elements.  `unique_mul A B a0 b0` asserts `a0 * b0` can be written in at
+most one way as a product of an element of `A` and an element of `B`. -/
+@[to_additive "Let `G` be a Type with addition, let `A B : finset G` be finite subsets and
+let `a0 b0 : G` be two elements.  `unique_add A B a0 b0` asserts `a0 * b0` can be written in at
+most one way as a sum of an element from `A` and an element from `B`."]
+def unique_mul {G} [has_mul G] (A B : finset G) (a0 b0 : G) : Prop :=
+∀ ⦃a b⦄, a ∈ A → b ∈ B → a * b = a0 * b0 → a = a0 ∧ b = b0
+
+namespace unique_mul
+variables {G H : Type*} [has_mul G] [has_mul H] {A B : finset G} {a0 b0 : G}
+
+lemma mt {G} [has_mul G] {A B : finset G} {a0 b0 : G} (h : unique_mul A B a0 b0) :
+  ∀ ⦃a b⦄, a ∈ A → b ∈ B → a ≠ a0 ∨ b ≠ b0 → a * b ≠ a0 * b0 :=
+λ _ _ ha hb k, by { contrapose! k, exact h ha hb k }
+
+@[to_additive]
+lemma subsingleton (A B : finset G) (a0 b0 : G) (h : unique_mul A B a0 b0) :
+  subsingleton { ab : G × G // ab.1 ∈ A ∧ ab.2 ∈ B ∧ ab.1 * ab.2 = a0 * b0 } :=
```

Comment:
Is it better to use `subsingleton` on a subtype or `set.subsingleton` on a set?

Reply:
I do not know: I added the `set.subsingleton`  lemma,  but I do not have an opinion of which one is better.